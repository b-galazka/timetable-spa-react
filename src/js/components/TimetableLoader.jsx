import React, {Component} from 'react';
import {connect} from 'react-redux';

import LoadingAnimation from './LoadingAnimation';

import errors from '../../json/errors';

import '../../scss/noContent.scss';

function mapStateToProps(state) {

    const {fetching, fetchingError, notExists} = state.timetableObject;

    return {

        timetableObject: {
            fetching,
            fetchingError,
            notExists
        }    
    };
}

class TimetableLoader extends Component {

    render() {

        return (
            <section className="timetable__loader">
                {this.renderLoaderContent()}
            </section>
        );
    }

    renderLoaderContent() {

        const {timetableObject} = this.props;

        if (timetableObject.fetching) {

            return <LoadingAnimation />;
        } else if (timetableObject.fetchingError) {

            return <p className="timetable__fetching-error">{errors.fetchingError}</p>;
        } else if (timetableObject.notExists) {

            return <p className="timetable__fetching-error">{errors.notFound}</p>;
        } else {

            return null;
        }
    }
}

export default connect(mapStateToProps)(TimetableLoader);