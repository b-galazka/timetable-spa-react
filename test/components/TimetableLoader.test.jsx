import React from 'react';
import renderer from 'react-test-renderer';

import TimetableLoader from '../../src/js/components/TimetableLoader';

jest.mock('../../src/js/components/LoadingAnimation', () => 'LoadingAnimation');

jest.mock(
    '../../src/json/errors',
    () => ({ fetchingError: 'fetching error', notFound: 'not found' })
);

describe('TimetableLoader component', () => {

    it('should render loading animation if data has not been fetched yet', () => {

        const tree = renderer.create(
            <TimetableLoader.WrappedComponent
                timetableObject={{
                    fetching: true,
                    fetchingError: false,
                    notExists: false
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render fetching error if it has occured during data fetching', () => {

        const tree = renderer.create(
            <TimetableLoader.WrappedComponent
                timetableObject={{
                    fetching: false,
                    fetchingError: true,
                    notExists: false
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render "not found" error if timetable object does not exist', () => {

        const tree = renderer.create(
            <TimetableLoader.WrappedComponent
                timetableObject={{
                    fetching: false,
                    fetchingError: false,
                    notExists: true
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});