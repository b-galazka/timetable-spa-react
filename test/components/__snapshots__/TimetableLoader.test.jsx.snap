// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimetableLoader component should render "not found" error if timetable object does not exist 1`] = `
<section
  className="timetable__loader"
>
  <p
    className="timetable__fetching-error"
  >
    not found
  </p>
</section>
`;

exports[`TimetableLoader component should render fetching error if it has occured during data fetching 1`] = `
<section
  className="timetable__loader"
>
  <p
    className="timetable__fetching-error"
  >
    fetching error
  </p>
</section>
`;

exports[`TimetableLoader component should render loading animation if data has not been fetched yet 1`] = `
<section
  className="timetable__loader"
>
  <LoadingAnimation />
</section>
`;
